backfill/2021-01-07-gpu-active/README.md

```diff
@@ -0,0 +1,199 @@
+the correct project.
+
+```bash
+gcloud config set project moz-fx-data-backfill-7
+
+# calculate the date range in this ugly one liner
+dates=$(python3 -c 'from datetime import datetime as dt, timedelta; \
+  start=dt.fromisoformat("2020-07-04"); \
+  end=dt.fromisoformat("2020-08-21"); \
+  days=(end-start).days; \
+  print(" ".join([(start + timedelta(i)).isoformat()[:10] for i in range(days)]))')
+
+# NOTE: prune tables in the live dataset for efficiency, see note below
+./script/copy_deduplicate --project-id moz-fx-data-backfill-7 --dates $(echo $dates)
+```
+
+This query was slow (hours) because it required iterating over every table in
+the project for a period of ~50 days. In the future, it is advisable to prune
+all of the unnecessary tables before running the script. Before running
+shredder, we'll delete all of the empty stable tables.
+
+```bash
+./prune-empty-tables
+```
+
+Now we run shredder from the bigquery-etl root.
+
+```bash
+script/shredder_delete \
+  --billing-projects moz-fx-data-backfill-7 \
+  --source-project moz-fx-data-shared-prod \
+  --target-project moz-fx-data-backfill-7 \
+  --start_date 2020-06-01 \
+  --only 'telemetry_stable.*' \
+  --dry_run
+```
+
+This removes relevant rows from our final tables.
+
+```bash
+INFO:root:Scanned 515495784 bytes and deleted 1280 rows from moz-fx-data-backfill-7.telemetry_stable.crash_v4
+INFO:root:Scanned 35301644397 bytes and deleted 45159 rows from moz-fx-data-backfill-7.telemetry_stable.event_v4
+INFO:root:Scanned 1059770786 bytes and deleted 169 rows from moz-fx-data-backfill-7.telemetry_stable.first_shutdown_v4
+INFO:root:Scanned 286322673 bytes and deleted 2 rows from moz-fx-data-backfill-7.telemetry_stable.heartbeat_v4
+INFO:root:Scanned 134028021311 bytes and deleted 13872 rows from moz-fx-data-backfill-7.telemetry_stable.main_v4
+INFO:root:Scanned 2795691020 bytes and deleted 1071 rows from moz-fx-data-backfill-7.telemetry_stable.modules_v4
+INFO:root:Scanned 302643221 bytes and deleted 163 rows from moz-fx-data-backfill-7.telemetry_stable.new_profile_v4
+INFO:root:Scanned 1245911143 bytes and deleted 6477 rows from moz-fx-data-backfill-7.telemetry_stable.update_v4
+INFO:root:Scanned 286924248 bytes and deleted 10 rows from moz-fx-data-backfill-7.telemetry_stable.voice_v4
+INFO:root:Scanned 175822424583 and deleted 68203 rows in total
+```
+
+We will need append these results to each of the shared prod live tables
+(requires ops-level permissions):
+
+```bash
+# generate the commands to be run
+bq ls --format json --project_id=moz-fx-data-backfill-7 telemetry_stable | \
+  jq -r '.[] | .tableReference.tableId' | \
+  xargs -I{} echo bq cp --append_table \
+    moz-fx-data-backfill-7:telemetry_stable.{} \
+    moz-fx-data-shared-prod:telemetry_stable.{}
+```
+
+```bash
+# commands to be run
+bq cp --append_table moz-fx-data-backfill-7:telemetry_stable.crash_v4 moz-fx-data-shared-prod:telemetry_stable.crash_v4
```

Comment:
I just attempted this first command and got:

```
BigQuery error in cp operation: Error processing job 'bigquery-etl-integration-
test:bqjob_r3a314ebbdcdff1bf_00000176f72421ba_1': Failed to copy moz-fx-data-backfill-7:telemetry_stable.crash_v4 to
moz-fx-data-shared-prod:telemetry_stable.crash_v4: incompatible clustering fields.
```

Reply:
Two things we could do here. We could simply make `bq update` calls on all these tables to set the clustering to `normalized_channel,sample_id`, in which case the `bq cp` should then work, although the underlying files won't be optimized for the correct clustering.

The better thing to do would be to create new stable tables in another backfill project, run a query on each one to populate from these, and then we can do the `bq cp` from the new tables.

Reply:
Oops, I'll go that second route.